src/commands/registerCommands.ts

```diff
@@ -366,3 +367,82 @@ function registerRestartLanguageServerCommand(context: ExtensionContext, languag
 
   }));
 }
+
+interface SurroundWithResponse {
+  start: TextEdit;
+  end: TextEdit;
+}
+
+class SurroundWithKind {
+
+  static readonly tags= 'tags';
+  static readonly comments= 'comments';
+  static readonly cdata= 'cdata';
+
+}
+
+/**
+ * Register commands used for refactoring XML files
+ *
+ * @param context the extension context
+ */
+function registerRefactorCommands(context: ExtensionContext, languageClient: LanguageClient) {
+
+  // Surround with Tags (Wrap)
+  context.subscriptions.push(commands.registerCommand(ClientCommandConstants.REFACTOR_SURROUND_WITH_TAGS, async () => {
+    await surroundWith(SurroundWithKind.tags, languageClient);
+  }));
+
+  // Surround with Comments
+  context.subscriptions.push(commands.registerCommand(ClientCommandConstants.REFACTOR_SURROUND_WITH_COMMENTS, async () => {
+    await surroundWith(SurroundWithKind.comments, languageClient);
+  }));
+
+
+  // Surround with CDATA
+  context.subscriptions.push(commands.registerCommand(ClientCommandConstants.REFACTOR_SURROUND_WITH_CDATA, async () => {
+    await surroundWith(SurroundWithKind.cdata, languageClient);
+  }));
+}
+
+async function surroundWith(surroundWithType: SurroundWithKind, languageClient: LanguageClient) {
+  const editor = window.activeTextEditor;
+  if (!editor) {
+    return;
+  }
+  const selection = editor.selections[0];
+  if (!selection) {
+    return;
+  }
+
+  const uri = window.activeTextEditor.document.uri;
+  const identifier = TextDocumentIdentifier.create(uri.toString());
+  const range = languageClient.code2ProtocolConverter.asRange(selection);
+  // Set the custom condition to watch if file already has bound grammar
+  let result: SurroundWithResponse;
+  try {
+    result = await commands.executeCommand(ServerCommandConstants.REFACTOR_SURROUND_WITH, identifier, range, surroundWithType);
+  } catch (error) {
+    console.log(`Error while surround with : ${error}`);
+  }
+
+  const startTag = result.start.newText;
+  const endTag = result.end.newText;
+  const startPos = languageClient.protocol2CodeConverter.asPosition(result.start.range.start);
+  const endPos = languageClient.protocol2CodeConverter.asPosition(result.end.range.start);
+  const pos = new Position(selection.start.line, selection.start.character + 1);
+
+  editor.insertSnippet(new SnippetString(endTag), endPos);
+  editor.insertSnippet(new SnippetString(startTag), startPos);
```

Comment:
If you apply these separately, then you have to undo twice to undo surround tag. I think it would be very helpful if we could avoid this.

Reply:
Sorry it was a POC, now I'm using SnippetTextEdit which works with WorkspaceEdit.